### ISSUE TITLE
  TypeError: Cannot read property 'internal' of null

### ISSUE THREAD

**Issue by trickydisco78** [datocms/gatsby-source-datocms#11]
Getting an error when running gatsby develop

node 8.94
yarn 1.3.2

Ran the gatsby new command and added the gatsby-source-datocms plugin. Filled out the files in gatsby-config.js and getting this when running gatsby develop

![image](https://user-images.githubusercontent.com/3627429/36600949-b34aba9e-18ab-11e8-8dff-87f508ad2391.png)

Am i missing something?

I've just encountered the same thing after cloning the gatsby-portfolio code and adding the datocms api key

**Comment by trickydisco78** [comment]
I can get the gatsby portfolio site to work. How do i get my own site to work. Do i need to add graphql queries or will it work without them?

**Comment by thetre97** [comment]
I have the same issue - and I can't even get the portfolio site to work?

```bash
error UNHANDLED REJECTION


  TypeError: Cannot read property 'internal' of null

  - infer-graphql-input-fields.js:222
    [orcatecdesignv2]/[gatsby]/dist/schema/infer-graphql-input-fields.js:222:38

  - lodash.js:4917
    [orcatecdesignv2]/[lodash]/lodash.js:4917:15

  - lodash.js:3002 baseForOwn
    [orcatecdesignv2]/[lodash]/lodash.js:3002:24

  - lodash.js:4886
    [orcatecdesignv2]/[lodash]/lodash.js:4886:18

  - lodash.js:9334 Function.forEach
    [orcatecdesignv2]/[lodash]/lodash.js:9334:14

  - infer-graphql-input-fields.js:209 inferInputObjectStructureFromNodes
    [orcatecdesignv2]/[gatsby]/dist/schema/infer-graphql-input-fields.js:209:5

  - build-node-types.js:88 _callee$
    [orcatecdesignv2]/[gatsby]/dist/schema/build-node-types.js:88:36

  - next_tick.js:150 process._tickCallback
    internal/process/next_tick.js:150:11
```

**Comment by thetre97** [comment]
I have just tried again.. and it all seems to work okay now? Can get both my website and the demo portfolio site to work.

**Comment by patricknick** [comment]
I ran into the same error message today with my website project. I noticed it must have to do something with the model I configured in DatoCMS and the created content.

In my models I defined a model which was allowed to be sorted in a tree-like structure:
<img width="300" alt="screenshot" src="https://user-images.githubusercontent.com/25477019/37023930-ccbe2a22-2127-11e8-8b6b-77eedc7271cf.png">

With this setup I received the same error as described in the other posts.

```
success source and transform nodes — 0.427 s
error UNHANDLED REJECTION


  TypeError: Cannot read property 'internal' of null

  - infer-graphql-input-fields.js:222
    [sample]/[gatsby]/dist/schema/infer-graphql-input-fields.js:222:37

  - lodash.js:4917
    [sample]/[lodash]/lodash.js:4917:15

  - lodash.js:3002 baseForOwn
    [sample]/[lodash]/lodash.js:3002:24

  - lodash.js:4886
    [sample]/[lodash]/lodash.js:4886:18

  - lodash.js:9334 Function.forEach
    [sample]/[lodash]/lodash.js:9334:14

  - infer-graphql-input-fields.js:209 inferInputObjectStructureFromNodes
    [sample]/[gatsby]/dist/schema/infer-graphql-input-fields.js:209:5

  - build-node-types.js:88 _callee$
    [sample]/[gatsby]/dist/schema/build-node-types.js:88:36

  - next_tick.js:131 _combinedTickCallback
    internal/process/next_tick.js:131:7


error Command failed with exit code 1.
```


Found out you need to have **at least 2 instances of this model** (by defining the in the content section of DatoCMS). If you do that, it should work.

**Comment by trickydisco78** [comment]
I also found it was something to do with the space /project i was using on dato. not sure why as it was very simple using 2 fields. Changed to a new one and everything worked

**Comment by thetre97** [comment]
I previously changed to a new space on datocms, and it worked again as @trickydisco78  said. However, I am now on a paid space, with all the models configured etc. I tried creating more than two instances, as @patricknick suggested, but that doesn't seem to work for me :( .

Anyone else have any suggestions? Would be grateful for any help.

**Comment by thetre97** [comment]
So further to that, I deleted the last model I had created, and removed any references to it in Gatsby. It worked! Sweet.

So then I recreated the model in Dato, and ran gatsby develop again, *without* adding any reference or gql queries to the Gatsby code. It threw this back at me?

```bash
error Plugin gatsby-source-datocms returned an error


  TypeError: Cannot read property 'children' of undefined

  - actions.js:409
    [orcatecdesign]/[gatsby]/dist/redux/actions.js:409:46

  - Array.map

  - actions.js:408 actions.deleteNodes
    [orcatecdesign]/[gatsby]/dist/redux/actions.js:408:41

  - bindActionCreators.js:7
    [orcatecdesign]/[redux]/lib/bindActionCreators.js:7:35

  - api-runner-node.js:47 doubleBoundActionCreators.(anonymous function)
    [orcatecdesign]/[gatsby]/dist/utils/api-runner-node.js:47:13

  - gatsby-node.js:48 _callee$
    [orcatecdesign]/[gatsby-source-datocms]/gatsby-node.js:48:27

  - gatsby-node.js:3 step
    [orcatecdesign]/[gatsby-source-datocms]/gatsby-node.js:3:191

  - gatsby-node.js:3
    [orcatecdesign]/[gatsby-source-datocms]/gatsby-node.js:3:437

  - es6-promise.js:458 initializePromise
    [orcatecdesign]/[es6-promise]/dist/es6-promise.js:458:5

  - es6-promise.js:874 new Promise
    [orcatecdesign]/[es6-promise]/dist/es6-promise.js:874:33

  - gatsby-node.js:3
    [orcatecdesign]/[gatsby-source-datocms]/gatsby-node.js:3:99

  - gatsby-node.js:84 sync
    [orcatecdesign]/[gatsby-source-datocms]/gatsby-node.js:84:30

  - gatsby-node.js:89 _callee2$
    [orcatecdesign]/[gatsby-source-datocms]/gatsby-node.js:89:20


warning The gatsby-source-datocms plugin has generated no Gatsby nodes. Do you need it?
```

This is just from adding an extra model called projects, with a single title field, and no instances. and to say again, with *no* reference to that model in Gatsby.

**Comment by patricknick** [comment]
The `TypeError: Cannot read property 'children' of undefined` error is not related to any reference in Gatsby. As far as I understand, this error happens before anything is done in your app code. The gatsby-source-datocms plugin tries to create an internal model based on the model and data on your CMS instance. That's why you receive this error even though you have no reference to the according models. 

I couldn't figure out the specific reason for the `TypeError: Cannot read property 'children' of undefined`. Last time I saw this, it happened because my model had defined a modular content block ([Modular Content](https://www.datocms.com/docs/introduction/modular-content/)) and in my model instances I didn't define any content yet. @thetre97 Probably the reason for your error is a similar one? Like no instance with an actual title defined? @stefanoverna Is this an edge case that needs to be caught more gracefully in the plugin?

.cache is the folder where the Gatsby and it's plugins cache the models and content. Couldn't find any documentation on this, so this is just my guess. :) But deleting it can help, yes.

**Comment by dardub** [comment]
I ran into this issue as well. I had a project with a few fields which included Modular Content. In an attempt to narrow the cause, I removed every field one by one, and the error never went away.

So I created a new project with 1 text field and one record. It did not throw the error.

I am able to reproduce it when I have a Modular Content field with no content in that record. By adding content the error goes away. Removing it, the error returns.

I'm just not sure what happened with the first project. With my first project, I could not get the error to go away at all.

**Comment by mwickett** [comment]
I'm seeing this issue as well, and reproduced it with a fresh Gatsby project pointing to the same Dato instance, which seems to me to make it clear that it's something wrong on the Dato project side.

This recently closed GatsbyJS issue discusses the same issue with Contentful and linked fields: https://github.com/gatsbyjs/gatsby/issues/4479

@stefanoverna - any insight into this issue? Thank you!

**Comment by stefanoverna** [comment]
Yes, unfortunately the problem lies in how Gatsby tries to infer GraphQL types from plugins. We can't do much about this issue until Gatsby offers some alternative API to plugins. Basically, for the plugin to work, right now you need to have at least one record for every model in your site, otherwise Gatsby will complain 😢

**Comment by stefanoverna** [comment]
This is the official issue on Gatsby: https://github.com/gatsbyjs/gatsby/issues/3344

**Comment by igghera** [comment]
> Small update: gatsby@1.9.260 seems to solve (sometimes) the problem

This worked for me, thanks!

**Comment by matjack1** [comment]
@stefanoverna have you seen this: https://github.com/gatsbyjs/gatsby/issues/10856 ?

**Comment by humbertqz** [comment]
> So, apparently, deleting `node_modules` & `.cache` folders, and reinstalling all packages again, then creating another model seems to set everything back to rights...

This worked for me too.

**Comment by stefanoverna** [comment]
Should be FINALLY solved in version `2.1.0-beta.1`, soon to be released! And please upgrade to the latest Gatsby version!

https://github.com/datocms/gatsby-source-datocms/blob/2.1.0/CHANGELOG.md#v210
